src/ccl_core.c

```diff
@@ -0,0 +1,786 @@
+    params->Omega_n_mass = ccl_Omeganuh2(1.0, params->N_nu_mass, params->mnu, params->T_CMB, NULL, status) / ((params->h)*(params->h));
+    ccl_check_status_nocosmo(status);
+  } 
+  else{
+    params->Omega_n_mass = 0.;
+  }
+  
+  params->Omega_m = params->Omega_b + params-> Omega_c;
+  params->Omega_l = 1.0 - params->Omega_m - params->Omega_g - params->Omega_n_rel -params->Omega_n_mass- params->Omega_k;
+  // Initially undetermined parameters - set to nan to trigger
+  // problems if they are mistakenly used.
+  if (isfinite(params->A_s)) {params->sigma8 = NAN;}
+  if (isfinite(params->sigma8)) {params->A_s = NAN;}
+  params->z_star = NAN;
+
+  if(fabs(params->Omega_k)<1E-6)
+    params->k_sign=0;
+  else if(params->Omega_k>0)
+    params->k_sign=-1;
+  else
+    params->k_sign=1;
+  params->sqrtk=sqrt(fabs(params->Omega_k))*params->h/CLIGHT_HMPC;
+}
+
+
+/* ------ ROUTINE: ccl_parameters_create -------
+INPUT: numbers for the basic cosmological parameters needed by CCL
+TASK: fill params with some initial values provided by the user
+DEFINITIONS:
+Omega_c: cold dark matter
+Omega_b: baryons
+Omega_m: matter
+Omega_k: curvature
+little omega_x means Omega_x*h^2
+Neff : Effective number of neutrino speces
+mnu : Pointer to either sum of neutrino masses or list of three masses.
+mnu_type : how the neutrino mass(es) should be treated
+w0: Dark energy eq of state parameter
+wa: Dark energy eq of state parameter, time variation
+H0: Hubble's constant in km/s/Mpc.
+h: Hubble's constant divided by (100 km/s/Mpc).
+A_s: amplitude of the primordial PS
+n_s: index of the primordial PS
+
+ */
+ccl_parameters ccl_parameters_create(
+                     double Omega_c, double Omega_b, double Omega_k,
+				     double Neff, double* mnu, ccl_mnu_convention mnu_type,
+				     double w0, double wa, double h, double norm_pk,
+				     double n_s, double bcm_log10Mc, double bcm_etab, 
+				     double bcm_ks, int nz_mgrowth, double *zarr_mgrowth,
+				     double *dfarr_mgrowth, int *status)
+{
+  #ifndef USE_GSL_ERROR
+    gsl_set_error_handler_off ();
+  #endif
+
+  ccl_parameters params;
+  // Initialize params
+  params.mnu = NULL;
+  params.z_mgrowth=NULL;
+  params.df_mgrowth=NULL;
+  params.sigma8 = NAN;
+  params.A_s = NAN;
+  params.Omega_c = Omega_c;
+  params.Omega_b = Omega_b;
+  params.Omega_k = Omega_k;
+  params.Neff = Neff;
+  
+  // Set the sum of neutrino masses
+  params.sum_nu_masses = *mnu;
+  double mnusum = *mnu;
+  double *mnu_in = NULL;
+  
+  /* Check whether ccl_splines and ccl_gsl exist. If either is not set yet, load
+     parameters from the config file. */
+  if(ccl_splines==NULL || ccl_gsl==NULL) {
+    ccl_cosmology_read_config();
+  }
+  /* Exit gracefully if config file can't be opened. */
+  if(ccl_splines==NULL || ccl_gsl==NULL) {
+    ccl_raise_exception(CCL_ERROR_MISSING_CONFIG_FILE, "ccl_core.c: Failed to read config file.");
+    *status = CCL_ERROR_MISSING_CONFIG_FILE;
```

Comment:
> If I followed the code correctly, this error only prints out if the config file can't be opened. Not if there was an error reading it. But then ccl_cosmology_read_config should already have "raised" an exception, so this is redundant.

This doesn't seem quite right. The reading routine doesn't check for memory allocation errors. This would mean the code would die anyways, but I get why the extra check is there. I am going to clean up the error handling, removing the one above and adding checks for memory.

>But maybe more importantly, I think this config file should really be read in at the python layer and passed as a struct to C. Then the user has the ability to change some of the parameters to suit their particular use case (e.g. more speed at the expense of some accuracy)

I agree with you but this has larger API implications. Right now the CCL team is supporting both a C and a Python API. If we force the config to be read by the python and not the C, then we are effectively deprecating the C API. We could add the ability to read it in at the python layer I suppose. However this has some overlap with issue #504. If we associate the data in this config with each cosmology, as opposed to having it be global, then we can support setting it via the python even if the config is read from C. I think for now this is the best path forward. It allows us to wait on deciding whether or not to deprecate the C API, cleans up some global state, and makes the library more configurable. 

Given this, I am going to punt on the second thing in favor of #504.